src/query/functions/src/scalars/decimal/src/cast.rs

```diff
@@ -379,13 +389,113 @@ pub fn convert_to_decimal(
+                    variant_to_decimal::<T>(arg, ctx, dest_type)
+                }
                 _ => unreachable!("to_decimal not support this DataType"),
             };
             result.upcast_decimal(size)
         }
     })
 }
 
+fn json_number_to_decimal<T>(
+    value: jsonb::Number,
+    dest_type: DecimalDataType,
+    max: T,
+    min: T,
+    multiplier: T,
+    multiplier_f64: f64,
+    ctx: &mut EvalContext,
+) -> Result<T, u32>
+where
+    T: Decimal + Mul<Output = T> + Div<Output = T>,
+{
+    let dest_size = dest_type.size();
+    match value {
+        jsonb::Number::Int64(v) => {
+            integer_to_decimal_scalar::<T, NumberType<i64>>(v, dest_size, multiplier)
+        }
+        jsonb::Number::UInt64(v) => {
+            integer_to_decimal_scalar::<T, NumberType<u64>>(v, dest_size, multiplier)
+        }
+        jsonb::Number::Float64(v) => {
+            let mut x = v * multiplier_f64;
+            if ctx.func_ctx.rounding_mode {
+                x = x.round();
+            }
+            let x = T::from_float(x);
+            if x > max || x < min {
+                Err(line!())
+            } else {
+                Ok(x)
+            }
+        }
+        jsonb::Number::Decimal128(d) => {
+            let from_size = DecimalSize {
+                precision: d.precision,
+                scale: d.scale,
+            };
+            let x = T::from_i128(d.value);
+            decimal_resize::<T>(ctx, x, from_size, dest_size, min, max)
+        }
+        jsonb::Number::Decimal256(d) => {
+            let from_size = DecimalSize {
+                precision: d.precision,
+                scale: d.scale,
+            };
+            match dest_type {
+                DecimalDataType::Decimal128(_) => {
+                    let x = i256(d.value);
+                    let min = i256::min_for_precision(dest_size.precision);
+                    let max = i256::max_for_precision(dest_size.precision);
+                    decimal_resize(ctx, x, from_size, dest_size, min, max).map(T::from_i256)
+                }
+                DecimalDataType::Decimal256(_) => {
+                    let x = T::from_i256(i256(d.value));
+                    decimal_resize::<T>(ctx, x, from_size, dest_size, min, max)
+                }
+            }
+        }
+    }
+}
+
+fn variant_to_decimal<T>(
+    from: Value<VariantType>,
+    ctx: &mut EvalContext,
+    dest_type: DecimalDataType,
+) -> Value<DecimalType<T>>
+where
+    T: Decimal + Mul<Output = T> + Div<Output = T>,
+{
+    let size = dest_type.size();
+    let multiplier = T::e(size.scale as u32);
+    let multiplier_f64: f64 = (10_f64).powi(size.scale as i32).as_();
+    let min = T::min_for_precision(size.precision);
+    let max = T::max_for_precision(size.precision);
+    let f = |val: &[u8], builder: &mut Vec<T>, ctx: &mut EvalContext| {
+        let raw_jsonb = RawJsonb::new(val);
+        let value = match raw_jsonb
+            .as_number()
```

Comment:
`variant` cast to `decimal` also support `boolean` type(`false` -> `0`, `true` -> `1`) and `string` type.
If the value is json `null`, convert it to `NULL`.

Reply:
bool and string supported